### ISSUE TITLE
Support all label types in DFToVW converter

### ISSUE THREAD

**Issue by jackgerrits** [VowpalWabbit/vowpal_wabbit#2452]
Currently only SimpleLabel is supported, it should eventually support all label types.

- [ ] multiclass
- [ ] mutlilabels
- [ ] cost sensitive (multiline)
- [ ] contextual bandit (multiline)
- [ ] ccb (multiline)
- [ ] slates (multiline)
- [ ] cb_eval (multiline)

**Comment by etiennekintzler** [comment]
Is the **multiclass** what's explained in [One Against All (oaa) multi class example](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/One-Against-All-(oaa)-multi-class-example)? If yes, I think the multiclass label type is quite straightforward to implement. The main difference with the `SimpleLabel` approach is that it required the label to be in the natural number set (between 1 and K). In fact `SimpleLabel` work just fine if the user convert the target column to natural number beforehand (but introducing `Multiclass` type might allow type checking and/or automatic conversion to natural numbers).

However I saw in the very wiki page that weights could be added to the target as well : 
> You may add weights to the example and the features, use name-spaces, etc.

Are weights the same thing as `Importance` in the [Input format wiki](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format) ? Since you said that the latter was under change should it be implemented ? Is the weight set by adding a space along with the desired weight (as in `1 0.3`) ?

Regarding **multilabels** is it what's explained in the [Cost Sensitive One Against All multi class example](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Cost-Sensitive-One-Against-All-%28csoaa%29-multi-class-example) ? If yes the main difficulty is that all the multiple labels can be present but not necessary, for instance : 
> 1:1.0 a1_expect_1| a
2:1.0 b1_expect_2| b
3:1.0 c1_expect_3| c
1:2.0 2:1.0 ab1_expect_2| a b
2:1.0 3:3.0 bc1_expect_2| b c
1:3.0 3:1.0 ac1_expect_3| a c
2:3.0 d1_expect_2| d

In your opinion, how would the input dataframe be like ?

**Comment by jackgerrits** [comment]
Yes, OAA uses multiclass labels. Weight is a part of this label as well as the class, it's effectively the same thing as importantance.
> Is the weight set by adding a space along with the desired weight (as in 1 0.3) ?

Yes

> Regarding multilabels is it what's explained in the Cost Sensitive One Against All multi class example ?

No, those use cost sensitive labels.

I'll try and get around to updating the input format wiki page soon to hopefully clear these things up.

**Comment by etiennekintzler** [comment]
Ok so csoaa is multi class as well ? I am confused since you mentionned it as a separate **multiline** type. 

To my understanding, the multi class pattern can be either : 
- just the label with optional importance (defined as `label importance`) as in [oaa](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/One-Against-All-(oaa)-multi-class-example) or [ect](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Error-Correcting-Tournament-(ect)-multi-class-example)
- one or more labels with optional cost (defined as `label:cost`) as in [wap](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Weighted-All-Pairs-(wap)-multi-class-example) or [csoaa](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Cost-Sensitive-One-Against-All-(csoaa)-multi-class-example)

For the case of multiple labels on the same line, how should the dataframe input be like ? For me the most natural would be to have as much columns as the maximal number of different label per line. For instance for the following vw ouput: 
> 1:1.0 a1_expect_1| a
2:1.0 b1_expect_2| b
3:1.0 c1_expect_3| c
1:2.0 2:1.0 ab1_expect_2| a b
2:1.0 3:3.0 bc1_expect_2| b c
1:3.0 3:1.0 ac1_expect_3| a c
2:3.0 d1_expect_2| d

the input dataframe would looks like
```python
   l1   c1   l2   c2
0   1  1.0  NaN  NaN
1   2  1.0  NaN  NaN
2   3  1.0  NaN  NaN
3   1  2.0  2  1.0
4   2  1.0  3  3.0
5   1  3.0  3  1.0
6   2  3.0  NaN  NaN
```

Also in the case that data arrive in a sparse format as for instance : 

```python
a = [dict(row=1, label=1, cost=1.0), dict(row=2, cost=2.0, label=1)]
b = [dict(row=2, label=2, cost=1.0)]
```
It is always possible to get back to a dataframe structure for instance by doing : 
```python
pd.merge(pd.DataFrame(a), pd.DataFrame(b), on="row", how="outer",
         suffixes=('_1', '_2')).set_index("row")
```
that yields : 
```python
     label_1  cost_1  label_2  cost_2
row                                  
1          1     1.0      NaN     NaN
2          1     2.0      2     1.0
```

What do you think ?

**Comment by jackgerrits** [comment]
I've updated the [input format wiki page](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format) to include more information about labels.

CSOAA (cost sensitive one against all) uses cost sensitive labels. Multiline is different to the type of label it expects. CSOAA has two variants, CSOAA (single line), CSOAA_LDF (multiline)

> just the label with optional importance (defined as label importance) as in oaa or ect

This is a mutliclass label

> one or more labels with optional cost (defined as label:cost) as in wap or csoaa

This is a cost sensitive label

The example dataframe looks good to me.

**Comment by JohnLangford** [comment]
Closing, but further patches are welcome.
